clonerefs: Use repo link from refs if available

taharah opened the pull request:
Dynamically sets the clone URI and path to flex based on the repo link from the ref if it is available. This helps further mitigate hard coded instances of "github.com" as noted on #14130. 

Signed-off-by: Trevor Wood <Trevor.G.Wood@gmail.com>

alvaroaleman commented on prow/pod-utils/clone/clone.go:
```diff
@@ -109,6 +110,9 @@ func PathForRefs(baseDir string, refs prowapi.Refs) string {
 		clonePath = refs.PathAlias
+	} else if refs.RepoLink != "" {
+		re := regexp.MustCompile(`^https?://`)
+		clonePath = re.ReplaceAllString(refs.RepoLink, "")
```
Do I understand correctly that the purpose of the regex is to strip the `http://` prefix? If so, `strings.TrimPrefix` would be more appropriate
/ok-to-test

taharah replied:
Yes, that is the intent; however, I used regex instead of `strings.TrimPrefix` in order to remove either `http://` or `https://` since more than likely there are still enterprise deployments without TLS enabled.

alvaroaleman replied:
Sure, so use two `string.Replace`s or just split by `://` and drop the first part? This problem seems a bit too trivial to use a regex

taharah commented:
@alvaroaleman I've addressed your comment. Additionally, just to provide some anecdotal evidence, I deployed my fork internally to our Prow cluster that uses GitHub Enterprise and the clonerefs utility worked exactly as expected.

Thank you for reviewing and happy holidays!

taharah commented:
/assign @stevekuznetsov

fejta commented on prow/pod-utils/clone/clone.go:
```diff
@@ -109,6 +109,10 @@ func PathForRefs(baseDir string, refs prowapi.Refs) string {
 	var clonePath string
 	if refs.PathAlias != "" {
 		clonePath = refs.PathAlias
+	} else if refs.RepoLink != "" {
```
In the future can we get these sort of changes behind flags? This broke gerrit integration (checking the repo out this way makes more sense, but is a backwards incompatible change that we need to be aware of)

alvaroaleman replied:
The issue is that very few ppl know what an incompatible change from gerrits POV is and those tend to not actively review changes. From githubs POV this is a backwards-compatible change.

fejta replied:
Change from "either use path_alias or else github.com/$ORG/$REPO" to now make repo_link take precedence is not a backwards compatible change.

alvaroaleman replied:
For github it is, because `RepoLink` without the protcol prefix and delimiter is the same as what is in the else block except for when it is a private GHE intstance, then it's the url to that instance and using that is more correct than always using github.com

fejta replied:
https://github.com/kubernetes/test-infra/blob/2abb7ff26579325f6a335990003f665755c96d7a/prow/gerrit/adapter/adapter.go#L154

fejta replied:
"I'm going to make this depend on RepoLink without a flag and also refuse to check whether anything depends on RepoLink" doesn't seem like a cooperative attitude

alvaroaleman replied:
This is not about attitude, this is about not being aware of how the gerrit integration uses things or what it even uses. I didn't even know it uses the decoration utils, so I didn't think about grepping anything, I just validated that its compatible from Githubs POV (which to all my knowledge it is).

Can you create a list of functionality that gerrit uses? I genuinely have no idea about that. I absolutely didn't expect a struct that is very clearly designed to hold data from Github to be used for something else.

fejta replied:
It uses very little in hook (aka plugins, etc) and basically all features of everything else.

fejta replied:
Is leaving a github comment unique to github? Yes, gerrit is unlikely to touch a comment call.

Is checking out a git repo unique to github? Not unique to github, something gerrit will likely depend on.

taharah replied:
@fejta I have to admit that I was also looking at this change solely from the perspective of Github and was not aware of the implications on the integrations of other git providers. Do you have an open issue tracking the impact of this change on Gerrit? I would like to help implement any fix(es) or solution that works for all providers.

fejta replied:
Thanks, and there's no issue to track since we just updated our paths to correct it.

This was a great contribution (thanks for making it!) and the correct solution. Nothing we need to change about the solution. We just needed to update our jobs to use the new path (which we've done).

Also I wouldn't expect you to be aware of these gotchas. Most people just use github, and core maintainers have been struggling to avoid these edge cases.

Ideally I'd like us to either:
* Be a little bit more diligent about checking the impacts of these changes
  - The new behavior changes RepoLink and ClonePath, which will gerrit depending on it
* Recognizing that this is often hard to, perhaps gating this behavior change behind a flag.
  - https://github.com/kubernetes/test-infra/pull/20339 is a good example of something which adds new functionality without immediately foisting it on everyone

fejta replied:
PS: the problem here is really subtle, so not super surprising to me that we missed it (both on impl, review, our deployment):
* It changes the checked out location for gerrit repos from `/root/github.com/somewhere.googlesource.com/bar` to `/root/somewhere.googlesource.com/bar`
  - The new path is obviously the more correct one
  - Not triggered if the ref specifies path_alias.


So if the job has problems if and only if it:
1) Uses gerrit
2) has gerrit refs that don't specify path_alias
3) Runs code that requires source code path to be /root/github.com/gerrit rather than /root/gerrit

As you might imagine most of our jobs fail at least one of these criteria -- they either aren't on gerrit, use path_alias and/or do not care so precisely where the code is located. Therefore it wasn't immediately obvious the new release was bad since most jobs were working.
